### PR TITLE
[cxx-interop] Fix linker errors when using std-vector.

### PR DESCRIPTION
**Pull request by zoecarver**
Adds tests for using std-vector and some other interesting types.

This patch fixes four mis conceptions that the compiler was previously making:
1. Implicit destructors have no side effects. (Yes, this means we were not cleaning up some objects.)
2. Implicit destructors have bodies. (Technically they do, but the body doesn't include CallExprs that they make when lowered to IR.)
3. Functions other than methods can be uninstantiated templates. 
4. Uninstantiated templates may have executable code. (I.e., we can never take the fast path.)

### CODE REVIEW

**Comment by zoecarver** [comment]
CC @plotfi

**Comment by zoecarver** [comment]
@guitard0g please verify this fixes the issues using `std::functional` and then add a test for that similar to my `std::vector` test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by plotfi** [comment]
@zoecarver Does this drop the need for PR #39551 ?

File: lib/IRGen/GenClangDecl.cpp [hunk 1]
```diff
@@ -100,7 +113,7 @@ void IRGenModule::emitClangDecl(const clang::Decl *decl) {
   SmallVector<const clang::Decl *, 8> stack;
   stack.push_back(decl);
 
-  ClangDeclFinder refFinder([&](const clang::Decl *D) {
+  auto callback = [&](const clang::Decl *D) {
     for (auto *DC = D->getDeclContext();; DC = DC->getParent()) {
```
Comment (plotfi) [inline lib/IRGen/GenClangDecl.cpp:117]: I believe this is where I was encountering the operator() for the std::functional that wasn't getting added to the stack.

File: lib/IRGen/GenClangDecl.cpp [hunk 2]
```diff
@@ -56,6 +57,17 @@ class ClangDeclFinder
     callback(CXXCE->getConstructor());
     return true;
   }
+
+  bool VisitVarDecl(clang::VarDecl *VD) {
+    if (auto cxxRecord = VD->getType()->getAsCXXRecordDecl())
+      if (auto dtor = cxxRecord->getDestructor())
+        callback(dtor);
```
Comment (plotfi) [inline lib/IRGen/GenClangDecl.cpp:64]: Ah I see, so this gets to the dtors from the VarDecl's type rather than the ctors it already visited?

**Comment by zoecarver** [comment]
> @zoecarver Does this drop the need for PR #39551 ?

Yes, I think so.

I'm sorry I wasn't able to review that patch more thoroughly, but I think this patch is a better solution. In that patch you are essentially walking the whole AST tree. It happens that your constructors lead you to the right place, but that is not guaranteed. I think many of these tests cases will fail with your patch, and it won't help the following use case: `func test(v: std.vector) { ... }` where we need to invoke the destructor on `v`, but we never call the constructor. This is why it's important that we look at the destructor for all `VarDecl`s (including parameters).

File: lib/IRGen/GenClangDecl.cpp (hunk 2 above)
Comment (zoecarver) [reply lib/IRGen/GenClangDecl.cpp:64]: Yes. The ctors aren't really important here; they weren't causing problems. We might need to destroy something that we never constructed. Think about it like this: what might cause us to call a destructor? Well, if we have an instance of a C++ type with a dtor, we will have to call its dtor. If that C++ type has bases/fields, same thing. To fix these issues, I look at all the places we might call a dtor, and make we tell clang to generate that symbol.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by plotfi** [comment]
> > @zoecarver Does this drop the need for PR #39551 ?
> 
> Yes, I think so.
> 
> I'm sorry I wasn't able to review that patch more thoroughly, but I think this patch is a better solution. In that patch you are essentially walking the whole AST tree. It happens that your constructors lead you to the right place, but that is not guaranteed. I think many of these tests cases will fail with your patch, and it won't help the following use case: `func test(v: std.vector) { ... }` where we need to invoke the destructor on `v`, but we never call the constructor. This is why it's important that we look at the destructor for all `VarDecl`s (including parameters).

Sounds good. I agree I like this patch a lot more, was going about tracking the Decls by walking from the ctors the entire depth is not great. I tested your patch out and it works too; will take a look at default template params for some of the other containers.

**Comment by plotfi** [comment]
> > @zoecarver Does this drop the need for PR #39551 ?
> 
> Yes, I think so.
> 
> I'm sorry I wasn't able to review that patch more thoroughly, but I think this patch is a better solution. In that patch you are essentially walking the whole AST tree. It happens that your constructors lead you to the right place, but that is not guaranteed. I think many of these tests cases will fail with your patch, and it won't help the following use case: `func test(v: std.vector) { ... }` where we need to invoke the destructor on `v`, but we never call the constructor. This is why it's important that we look at the destructor for all `VarDecl`s (including parameters).

No worries BTW! :-)

File: lib/IRGen/GenClangDecl.cpp (hunk 2 above)
Comment (plotfi) [reply lib/IRGen/GenClangDecl.cpp:64]: Good approach. Can you add this as context in a comment on this VarDecl visitor?

**Comment by zoecarver** [comment]
@swift-ci please smoke test.

**Comment by zoecarver** [comment]
@swift-ci please smoke test Linux.

**Pull request merged**